### DISCUSSION TITLE
destroying duplicate IKE_SA for peer

### DISCUSSION THREAD

**Discussion by trispace** [strongswan/strongswan#1000] [General]
I'd like to configure a redundant IPSec tunnel from spokes to a hub. Each spoke creates two IPSec tunnels to the hub and uses xfrm interfaces over which a routing protocol will be used. For authentication certificates will be used, each spoke will use a distinct certificate.

As seen in the log files, the initial connections from spoke1 to the hub works fine (two IKE_SAs and CHILD_SAs are created).

Unfortunately, after restarting the spoke, the hub logs `destroying duplicate IKE_SA for peer 'spoke1', received INITIAL_CONTACT` and destroys one of the two IKE_SAs and CHILD_SAs of from spoke1. I'm a little out of ideas how to fix this and would be thankful for any hints.

[swanctl-spoke1.txt](https://github.com/strongswan/strongswan/files/8518983/swanctl-spoke1.txt)
[swanctl-hub.txt](https://github.com/strongswan/strongswan/files/8518984/swanctl-hub.txt)
[strongswan-spoke1.log](https://github.com/strongswan/strongswan/files/8518985/strongswan-spoke1.log)
[strongswan-hub.log](https://github.com/strongswan/strongswan/files/8518987/strongswan-hub.log)

The strongswan version on both hub and spoke is 5.9.1 (Debian package 5.9.1-1+deb11u2)

**Comment by tobiasbrunner** [comment]
> Unfortunately, after restarting the spoke, the hub logs `destroying duplicate IKE_SA for peer 'spoke1', received INITIAL_CONTACT` and destroys one of the two IKE_SAs and CHILD_SAs of from spoke1.

While you could avoid that via `unique` option, the responder will always select the same config (i.e. it will always use if_id `0x100`). You either have to use distinct client identities for the two connections (you could use the same certificate if it contains both identities as subjectAltNames) or use a the new simple label option from the 5.9.6 release candidate to select a specific child config (as illustrated by the [ikev2/net2net-dscp scenario](https://www.strongswan.org/testing/testresults5rc/ikev2/net2net-dscp/index.html)).

**Reply by trispace** [reply]
Thank you very much for the swift reply!
Could it be that this behaviour changed in the last releases? I'm using a similar configuration for at least a year with version 5.8.4 where I never saw the message about the duplicate IKE_SAs. The only difference I could see is that in the working 5.8.4 configuration I'm using the full DN in the `id` option for matching the remote ID vs. the match on the SAN in the attached hub configuration file

**Reply by tobiasbrunner** [reply]
No, nothing changed. But if you use the DN for one connection and the SAN for the other, there will obviously not be any problem.

**Reply by trispace** [reply]
Both connections use the DN matching with version 5.8.4, therefore I wonder what the issue might be on 5.9.1.
But anyways, thank you very much for your time. I'll try to create another test setup and will report back if the duplicate IKE_SA issue happens again.

**Reply by tobiasbrunner** [reply]
You probably just didn't notice that it didn't do what you intended.

**Reply by trispace** [reply]
I'm pretty sure I would notice since the 5.8.4 setup is used with 23 spokes and both connections from each of them are established:
(connections with l1 as prefix are the spokes first connection, the ones with l2 as prefix the second one)

```
# swanctl --list-sas|grep ESTABLISHED|grep -c ^l1
23
# swanctl --list-sas|grep ESTABLISHED|grep -c ^l2
23
```

Another thing which I don't fully understand is why both initial SAs work fine until the reboot of the spoke (as seen in the attached log files). Wouldn't the hub detect and destroy the duplicate spoke1 IKE and CHILD_SA as soon as the second connection from spoke1 gets established initially?

**Reply by tobiasbrunner** [reply]
Ah, I just now noticed that you used different IP addresses. That actually ensures that the correct config is selected (even if the identities are the same). So that's fine, sorry for that.

That duplicates are not detected is probably due to a race condition. A duplicate is only found if there exists a fully-established IKE_SA with the same identities when processing an IKE_AUTH request (you see that difference in the log, initially the two requests are processed concurrently, while later on one is complete before the other is even received).

**Reply by trispace** [reply]
Thanks again for your patience!
To be honest, I suspected some kind of race condition. Maybe the race only happened on my 5.9.1 test setup, since it runs in containers on a Linux box, while the the 5.8.4 setup runs over the Internet with certainly higher packet delays between hub and spoke than on my test setup.
